src/main/java/minijava/ir/optimize/Inliner.java

```diff
@@ -32,27 +32,44 @@
 public class Inliner extends BaseOptimizer {
   private static final Set<ir_opcode> ALWAYS_IN_START_BLOCK =
       Sets.newHashSet(iro_Const, iro_Address);
+  private static final int MAX_NODES = 1000;
+  /**
+   * Specifies the maximal number of nodes a leaf method can have to always be inlined, regardless
+   * of whether the graph already has more than {@link #MAX_NODES} nodes or not.
+   */
+  private static final int MAX_LEAF_SIZE_TO_ALWAYS_INLINE = 80;
+
   private final ProgramMetrics metrics;
   private final Set<Call> callsToInline = new HashSet<>();
 
-  public Inliner(ProgramMetrics metrics) {
+  public final boolean onlyLeafs;
+
+  public Inliner(ProgramMetrics metrics, boolean onlyLeafs) {
     this.metrics = metrics;
+    this.onlyLeafs = onlyLeafs;
   }
 
   @Override
   public boolean optimize(Graph graph) {
     this.graph = graph;
     this.callsToInline.clear();
+    metrics.updateGraphInfo(graph);
     fixedPointIteration();
     return inlineCandidates();
   }
 
   private boolean inlineCandidates() {
     boolean hasChanged = false;
+    int size = metrics.graphInfos.get(graph).size;
     for (Call call : callsToInline) {
-      //System.out.println("Inlining " + callee(call) + " into " + graph);
-      inline(call);
-      hasChanged = true;
+      int methodSize = metrics.graphInfos.get(call.getGraph()).size;
+      int newGraphSize = methodSize + size;
+      if (onlyLeafs && (methodSize < MAX_LEAF_SIZE_TO_ALWAYS_INLINE || newGraphSize < MAX_NODES)
+          || (!onlyLeafs && newGraphSize < MAX_NODES)) {
```

Comment:
How about `onlyLeafs && methodSize < MAX_LEAF_SIZE_TO_ALWAYS_INLINE || newGraphSize < MAX_NODES`?

I have to truth table handy, but should be the same:

```
(a & (b | c)) | (!a & b)
= distribute a &
a & b | a& c | !a & b
= law of the excluded middle on a and !a
a& c | b
```

Otherwise, we are good!